igwn_auth_utils/requests.py

```diff
@@ -0,0 +1,121 @@
+# -*- coding: utf-8 -*-
+# Copyright 2021 Cardiff University
+# Distributed under the terms of the BSD-3-Clause license
+
+__author__ = "Duncan Macleod <duncan.macleod@ligo.org>"
+
+from functools import wraps
+
+import requests
+
+from . import (
+    find_x509_credentials,
+    find_scitoken,
+    scitoken_authorization_header,
+)
+
+
+def _nonfatal(func):
+    """Wrap a function to return `None` in the event of a `RuntimeError`
+    """
+    @wraps(func)
+    def wrapped(*args, **kwargs):
+        try:
+            return func(*args, **kwargs)
+        except RuntimeError:
+            return
+
+    return wrapped
+
+
+# non-fatal versions of the authorisation finders
+_find_x509_nonfatal = _nonfatal(find_x509_credentials)
+_find_scitoken_nonfatal = _nonfatal(find_scitoken)
+
+
+class SessionAuthMixin:
+    """Mixin for :class:`requests.Session` to add support for IGWN auth.
+
+    Parameters
+    ----------
+    cert : `str`, `tuple`, optional
+        path to a PEM-format certificate file, or a ``(cert, key)`` `tuple`;
+        if not given `igwn_auth_utils.find_x509_credentials` will be attempted.
+
+    token : `scitokens.SciToken`, optional
+        a authorisation bearer token; if not given
+        `igwn_auth_utils.find_scitoken` will be attempted.
+
+    token_audience, token_scope : `str`
+        the ``audience`` and ``scope`` to pass to
+        `igwn_auth_utils.find_scitoken` when discovering available tokens.
+
+    force_noauth : `bool`, optional
+        disable the use of any authorisation credentials (mainly for testing)
+        (default: `False`).
+
+    fail_if_noauth : `bool`, optional
+        raise an `Exception` if no authorisation credentials are presented
+        or discovered (default: `False`).
+
+    Raises
+    ------
+    ValueError
+        if ``fail_if_noauth=True`` is given and no authorisation
+        token/credentials are presented or discovered.
+    """
+    def __init__(
+        self,
+        cert=None,
+        token=None,
+        token_audience=None,
+        token_scope=None,
+        force_noauth=False,
+        fail_if_noauth=False,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+
+        if force_noauth and fail_if_noauth:
+            raise ValueError(
+                "cannot select both force_noauth and fail_if_noauth",
+            )
+        if force_noauth:
+            return
+
+        # handle X.509 certificate
+        self.cert = cert or _find_x509_nonfatal()
+
+        # handle bearer token
+        token = token or _find_scitoken_nonfatal(token_audience, token_scope)
+        if token:
+            self.headers.update({
+                "Authorization": scitoken_authorization_header(token),
+            })
+
+        # if no auth was found, and we need it, fail here
+        if not self._has_igwn_auth and fail_if_noauth:
+            raise ValueError("no valid authorisation credentials found")
```

Comment:
Does it make any sense for both the cert and the scitoken to be used in the same session?

I would suggest organizing this into a single `if/elif` block like I did so that it is easy to verify that exactly one authentication method is set up:

```python
        # FIXME: these should go into elif clauses below
        # (as in `elif default_cert := _find_x509_credentials():`)
        # in order to defer unnecessary I/O,  but this requires
        # the := operator, which requires Python 3.8.
        default_cert = _find_cert()
        default_basic_auth = _find_username_password(url)

        if force_noauth:
            pass
        elif cert is not None:
            self.cert = cert
        elif username is not None:
            self.auth = (username, password)
        elif default_cert is not None:
            self.cert = default_cert
        elif default_basic_auth is not None:
            self.auth = default_basic_auth
        elif fail_if_noauth:
            raise ValueError('No authentication credentials found.')
```